base/abstractarray.jl

```diff
@@ -2506,6 +2506,189 @@ end
+true
+
+julia> vec(mat) == reduce(vcat, vecs) == vcat(vecs...)
+true
+
+julia> mats = (fill(i/2,3,4) for i in (1, 10, 100) if i>pi);
+
+julia> stack(mats)
+3×4×2 Array{Float64, 3}:
+[:, :, 1] =
+ 5.0  5.0  5.0  5.0
+ 5.0  5.0  5.0  5.0
+ 5.0  5.0  5.0  5.0
+
+[:, :, 2] =
+ 50.0  50.0  50.0  50.0
+ 50.0  50.0  50.0  50.0
+ 50.0  50.0  50.0  50.0
+
+julia> ans == cat(mats..., dims=3)
+true
+```
+"""
+stack(itr) = _stack_iter(IteratorSize(itr), itr)
+stack(A::AbstractArray{<:AbstractArray}) = _typed_stack(mapreduce(eltype, promote_type, A), A)
+
+"""
+    stack(f, args)
+
+Apply `f` to each element of `args`, and `stack` the result.
+
+See also [`mapslices`](@ref), [`mapreduce`](@ref).
+
+# Examples
+```jldoctest
+julia> stack("julia") do c
+         (c, c-32)
+       end
+2×5 Matrix{Char}:
+ 'j'  'u'  'l'  'i'  'a'
+ 'J'  'U'  'L'  'I'  'A'
+
+julia> ans == mapreduce(c -> [c, c-32], hcat, "julia")
+true
+
+julia> stack(x -> x*x', eachcol([1 2; 10 20; 100 200]))
+3×3×2 Array{Int64, 3}:
+[:, :, 1] =
+   1    10    100
+  10   100   1000
+ 100  1000  10000
+
+[:, :, 2] =
+   4    40    400
+  40   400   4000
+ 400  4000  40000
+
+julia> ans == cat([1,10,100] * [1,10,100]', [2,20,200] * [2,20,200]'; dims=3)
+true
+```
+"""
+stack(f, itr) = stack(Iterators.map(f, itr))
+
+function _stack_iter(::HasShape, itr)
+    w, val = _vstack_plus(itr)
+    reshape(w, axes(val)..., axes(itr)...)
+end
+function _stack_iter(::IteratorSize, itr)
+    w, val = _vstack_plus(itr)
+    d = length(w) ÷ length(val)
+    reshape(w, axes(val)..., OneTo(d))
+end
+
+function _vstack_plus(itr)
+    z = iterate(itr)
+    z === nothing && throw(ArgumentError("cannot stack an empty collection"))
+    val, state = z
+    val isa Union{AbstractArray, Tuple} || throw(ArgumentError("cannot stack elements of type $(typeof(val))"))
+
+    axe = axes(val)
+    len = length(val)
+    n = haslength(itr) ? len*length(itr) : nothing
+
+    v = if val isa Tuple
+        T = mapreduce(typeof, promote_type, val)
+        similar(1:0, T, something(n, len))
+    else
+        similar(val, something(n, len))
+    end
+    copyto!(v, 1, val, firstindex(val), len)
+
+    w = _stack_rest!(v, 0, n, axe, itr, state)
+    w, val
+end
+
+function _stack_rest!(v::AbstractVector, i, n, axe, itr, state)
+    len = prod(length, axe; init=1)
+    while true
+        z = iterate(itr, state)
+        z === nothing && return v
+        val, state = z
+        axes(val) == axe || throw(DimensionMismatch(
+            "expected a consistent size, got axes $(UnitRange.(axes(val))) compared to $(UnitRange.(axe)) for the first"))
+        i += 1
+        T′ = if val isa Tuple
+            promote_type(eltype(v), mapreduce(typeof, promote_type, val))
+        else
+            promote_type(eltype(v), eltype(val))
+        end
+        if T′ <: eltype(v)
+            if n isa Integer
+                copyto!(v, i*len+1, val, firstindex(val), len)
+            else
+                append!(v, val)
+            end
```

Comment:
The `append!` path here is for when the length `n` isn't known. But trying to time things, I'm not really sure this is a good idea:
```julia
julia> v_v = [rand(128) for ii in 1:1000];

julia> @btime hcat($(v_v)...);
  min 94.834 μs, mean 155.820 μs (6 allocations, 1023.78 KiB. GC mean 11.29%)

julia> @btime reduce(hcat, $v_v);
  min 30.541 μs, mean 98.625 μs (2 allocations, 1000.05 KiB. GC mean 14.77%)

julia> @btime stack($v_v);  # array path
  min 27.166 μs, mean 105.319 μs (2 allocations, 1000.05 KiB. GC mean 19.79%)

julia> g_v = (x for x in v_v);

julia> @btime reduce(hcat, $g_v);  # (note the units!)
  min 34.717 ms, mean 39.499 ms (1983 allocations, 488.82 MiB. GC mean 22.92%)

julia> @btime stack($g_v);  # iterator path
  min 28.625 μs, mean 99.557 μs (4 allocations, 1000.14 KiB. GC mean 17.17%)

julia> @btime stack(collect($g_v));  # collecting instead
  min 27.708 μs, mean 104.952 μs (3 allocations, 1007.98 KiB. GC mean 20.16%)

julia> f_g_v = Iterators.filter(x->true, g_v);

julia> @btime reduce(hcat, $f_g_v);  # (note the units!)
  min 36.220 ms, mean 38.267 ms (1983 allocations, 488.82 MiB. GC mean 21.37%)

julia> @btime stack($f_g_v);  # uses append!
  min 72.125 μs, mean 185.355 μs (10 allocations, 1.82 MiB. GC mean 23.98%)

julia> @btime stack(collect($f_g_v));  # might always be better to collect?
  min 34.208 μs, mean 107.424 μs (9 allocations, 1021.94 KiB. GC mean 18.89%)
```
So it might always be faster to `collect` iterators of unknown length? This allocates an array to contain the slices, but this is probably small compared to the slices themselves, whose total size is the same as the final result. If we collect, all slices must be in memory simultaneously, can that lead to any cases where that's worse? 

Even for iterators of known length, the gain over collecting the iterator seems pretty small. 

Reading #31644 a bit more for examples, that also handles the case `reduce(vcat, vector_of_vectors)` with uneven lengths, which `stack` does not. Whether in that case `append!` beats calculating the length first, I don't know.

Reply:
An example where collecting is worse than repeated `append!`, but not by much:
```julia
julia> xs = (rand(rand(1:3)) for _ in 1:1000 if true);

julia> @btime pusher($xs);
  min 39.000 μs, mean 47.298 μs (1008 allocations, 125.94 KiB)

julia> @btime reduce(vcat, collect($xs));
  min 45.875 μs, mean 55.630 μs (1009 allocations, 116.89 KiB)
```

Reply:
New implementation removes the `append!` path completely. Iterators of unknown length will always be collected before proceeding.